### ISSUE TITLE
Kustomize build ComparisonError (unable to make loader)

### ISSUE THREAD

**Issue by pepe9012** [argoproj/argo-cd#15057]
**Describe the bug**
App templated with kustomize ends in error state (ComparisonError)

We encountered problem with kustomize (v4.5.7 and v5.1.1) and argocd **v2.8.0**. 
First, restarting `argocd-repo-server` solved the problem. However, the error reappeared and the restart did not fix it.
I did not find anything useful in the `debug` logs. With argocd v2.7.x everything worked fine.

Did not help:
- disabling redis compression
- update kustomize to v5
- restarting controllers

**To Reproduce**

The problem appears with kustomize templated apps. I have not been able to find the key to reproduce accurately, but with a `hard refresh` action, the error should appear. But it depends on how long the repo-server is running. 

In my case, app ends in the error state after hard-refresh triggered between 15-60minutes after repo-server restart.
Once kustomize build get `permission denied`, it never recovered without restarting repo-server.
 
**Expected behavior**

Kustomize apps do not fall in the error state after (hard)refresh.

**Version**

```shell
v2.8.0+804d4b8
```

**Logs**

```
Failed to load target state: failed to generate manifest for source 1 of 1: rpc error: code = Unknown desc = Manifest generation error (cached): `kustomize build <path to cached source>/path/to/my/app` failed exit status 1: 2023/08/15 12:31:20 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied
```

**Comment by praveen-livspace** [comment]
I thought I am the only one getting this error. For now the only resolution I know is to restart the repo server and hard refresh the app.

**Comment by pepe9012** [comment]
> I thought I am the only one getting this error. For now the only resolution I know is to restart the repo server and hard refresh the app.

yes it help, but not permanently.. i think that the problem will appear soon again so I will check long-term debug logs on repo-server and hopefully find some hint

**Comment by pepe9012** [comment]
finally, we upgraded a dozen of argocd instances, with 3000+ kustomize & helm apps, and only one app (in default project, which wasnt modified) have this problem... the app is deployed on all instances, but problem occurred only on two of them. its really like hunting a ghost

**Comment by pepe9012** [comment]
One of our instances with 1k applications is starting to fail with the expiration of the `repo cache`. The Kustomize build always ends with a `permission denied` error message and all applications go from the `Synced` sync state to the `Unknown` state.

So with `reposerver.repo.cache.expiration: "30m"`, kustomize build stop working every 30minutes.
Can be fixed in two ways:
1) restart redis cache and when its running again, restart all other components 
2) restart repo-server and force a hard refresh on all apps in Unknow sync state

**Comment by praveen-livspace** [comment]
Argocd version: v2.8.4+c279299
Applications randomly enter this sync state. Sometimes it resolves on hard refresh. But that too is non deterministic.

<img width="1277" alt="image" src="https://github.com/argoproj/argo-cd/assets/77090791/cfc259c7-49a7-4977-b974-22cf2a48d7f1">
<img width="1198" alt="image" src="https://github.com/argoproj/argo-cd/assets/77090791/becf6336-70b4-4ad3-b770-78242b7f091e">

**Comment by crenshaw-dev** [comment]
I believe the error is probably from here: https://github.com/kubernetes-sigs/kustomize/blob/7e541ca380863ccdfc05db2ef8b97e9f0ccff2d6/kyaml/filesys/fsondisk.go#L56

Argo CD repo-server zeroes out filesystem permissions for a cached repo (the path represented in the error as `<path to cached source>`) when the source is not in active use.

My guess is that there's some race condition causing the permissions to be zeroed out before Kustomize is able to finish its work. 

Inspecting repo-server logs around the time of this error may give some indication as to what process locked the directory before the Kustomize process completed.

**Comment by praveen-livspace** [comment]
```
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=debug msg="Successfully removed permissions on /tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c"
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg=Trace args="[kustomize build /tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c/static/production/kamar-taj --load-restrictor LoadRestrictionsNone --enable-helm]" dir= operation_name="exec kustomize" time_ms=3.375581
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=error msg="`kustomize build /tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c/static/production/kamar-taj --load-restrictor LoadRestrictionsNone --enable-helm` failed exit status 1: 2023/10/13 06:00:18 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied" execID=cbef8
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg="kustomize build /tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c/static/production/kamar-taj --load-restrictor LoadRestrictionsNone --enable-helm" dir= execID=cbef8
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg=Trace args="[kustomize edit set namespace -- prod]" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c/static/production/kamar-taj operation_name="exec kustomize" time_ms=7.265429
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg="kustomize edit set namespace -- prod" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c/static/production/kamar-taj execID=1d91d
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg=Trace args="[kustomize edit add annotation part-of:kamar-taj team:sales-and-growth]" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c/static/production/kamar-taj operation_name="exec kustomize" time_ms=7.3034930000000005
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg="kustomize edit add annotation part-of:kamar-taj team:sales-and-growth" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c/static/production/kamar-taj execID=2296c
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg=Trace args="[git rev-parse HEAD]" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c operation_name="exec git" time_ms=0.9256260000000001
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg="git rev-parse HEAD" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c execID=8aa30
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg=Trace args="[git clean -ffdx]" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c operation_name="exec git" time_ms=3.307339
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg="git clean -ffdx" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c execID=189e8
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg=Trace args="[git checkout --force 169b66260f78251e6dc76a7984adeccfa66bb29c]" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c operation_name="exec git" time_ms=3.9300439999999996
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg="git checkout --force 169b66260f78251e6dc76a7984adeccfa66bb29c" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c execID=40b06
2023-10-13T11:30:18+05:30	time="2023-10-13T06:00:18Z" level=info msg=Trace args="[git fetch origin --tags --force --prune]" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c operation_name="exec git" time_ms=1149.534914
2023-10-13T11:30:17+05:30	time="2023-10-13T06:00:17Z" level=info msg="git fetch origin --tags --force --prune" dir=/tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c execID=be1d9
2023-10-13T11:30:17+05:30	time="2023-10-13T06:00:17Z" level=debug msg="Successfully restored read/write/execute permissions on /tmp/_argocd-repo/e3978565-3c7f-45fc-b070-e3cc1ce9f44c"
```
Sample log lines from the repo server.

**Comment by prune998** [comment]
I also had this issue with 2 apps using the same repo. 
As explained, restarting the `repo-server` and doing a `hard refresh` of the app resolved the issue.

Before restarting the `repo-server`, I also had a new `App` created from an `AppSet` that wasn't displayed in the UI... my guess is that the whole `repo-server` was bad behaving ?

**Comment by pepe9012** [comment]
@crenshaw-dev Here is the full log where I first performed a "hard-refresh" at 7:58:34. It looks like the permissions are removed after the 'kustomize build' process, but for some reason it fails.

For testing purposes, I reduced the number of repo-server replicas to 1 and disabled parallel processes (`reposerver.parallelism.limit: 1`).

We use gitlab enterprise on a private domain. The ArgoCD instance has hundreds of 'Kustomize' applications spread over several projects and git repsitories. Applications get into an "Unknown" state and only restarting the repo-server will help. But not for long. 

Do you have any idea, where we should look for more info?

```
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="symbolic reference 'HEAD' (refs/heads/main) resolved to 'a3fffe6133d7300a5b89daf639c7f13ab50ba9fc'"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="cache hit for repo: https://gitlab.my-domain.com/my-repo.git revision: a3fffe6133d7300a5b89daf639c7f13ab50ba9fc"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=GetGitDirectories grpc.service=repository.RepoServerService grpc.start_time="2023-11-03T07:58:34Z" grpc.time_ms=1.046 span.kind=server system=grpc
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="symbolic reference 'HEAD' (refs/heads/main) resolved to 'a3fffe6133d7300a5b89daf639c7f13ab50ba9fc'"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="Successfully restored read/write/execute permissions on /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg="git fetch origin --tags --force --prune" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 execID=8a1f0
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="symbolic reference 'HEAD' (refs/heads/main) resolved to 'a3fffe6133d7300a5b89daf639c7f13ab50ba9fc'"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug duration=535.648263ms execID=8a1f0
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg=Trace args="[git fetch origin --tags --force --prune]" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 operation_name="exec git" time_ms=535.753422
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg="git checkout --force a3fffe6133d7300a5b89daf639c7f13ab50ba9fc" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 execID=015bd
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug duration=4.121009ms execID=015bd
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg=Trace args="[git checkout --force a3fffe6133d7300a5b89daf639c7f13ab50ba9fc]" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 operation_name="exec git" time_ms=4.196693
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg="git clean -ffdx" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 execID=17582
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug duration=2.568751ms execID=17582
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg=Trace args="[git clean -ffdx]" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 operation_name="exec git" time_ms=2.642462
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg="git rev-parse HEAD" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 execID=84c26
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg="git rev-parse HEAD" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 execID=2cb32
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="a3fffe6133d7300a5b89daf639c7f13ab50ba9fc\n" duration=2.630664ms execID=84c26
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg=Trace args="[git rev-parse HEAD]" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 operation_name="exec git" time_ms=2.723641
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="a3fffe6133d7300a5b89daf639c7f13ab50ba9fc\n" duration=2.479433ms execID=2cb32
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg=Trace args="[git rev-parse HEAD]" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 operation_name="exec git" time_ms=2.563066
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="pluginSockFilePath is: /home/argocd/cmp-server/plugins" CWE=775 security=1
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="pluginSockFilePath is: /home/argocd/cmp-server/plugins" CWE=775 security=1
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="No caCert found for repo gitlab.my-domain.com"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg="kustomize build /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784/components/my-app/staging" dir= execID=bb343
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="No caCert found for repo gitlab.my-domain.com"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg="kustomize build /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784/components/my-app/staging" dir= execID=5ad36
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug duration=8.934628ms execID=bb343
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=error msg="`kustomize build /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784/components/my-app/staging` failed exit status 1: 2023/11/03 07:58:34 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied" execID=bb343
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg=Trace args="[kustomize build /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784/components/my-app/staging]" dir= operation_name="exec kustomize" time_ms=9.098927
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="getting manifests cache" appName=my-app-staging appSrc="{\"appSrc\":{\"repoURL\":\"\",\"path\":\"components/my-app/staging\"},\"srcRefs\":{}}" clusterInfo="1.22|<<edited>>" namespace=staging reason="GenerateManifests error" revision=a3fffe6133d7300a5b89daf639c7f13ab50ba9fc trackingKey="annotation:app.kubernetes.io/instance"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug duration=9.902219ms execID=5ad36
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=error msg="`kustomize build /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784/components/my-app/staging` failed exit status 1: 2023/11/03 07:58:34 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied" execID=5ad36
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg=Trace args="[kustomize build /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784/components/my-app/staging]" dir= operation_name="exec kustomize" time_ms=10.068526
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="setting manifests cache" appName=my-app-staging appSrc="{\"appSrc\":{\"repoURL\":\"\",\"path\":\"components/my-app/staging\"},\"srcRefs\":{}}" clusterInfo="1.22|<<edited>>" namespace=staging reason="GenerateManifests error" revision=a3fffe6133d7300a5b89daf639c7f13ab50ba9fc trackingKey="annotation:app.kubernetes.io/instance"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=error msg="finished unary call with code Unknown" error="`kustomize build <path to cached source>/components/my-app/staging` failed exit status 1: 2023/11/03 07:58:34 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied" grpc.code=Unknown grpc.method=GetAppDetails grpc.service=repository.RepoServerService grpc.start_time="2023-11-03T07:58:34Z" grpc.time_ms=650.131 span.kind=server system=grpc
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="Successfully removed permissions on /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=error msg="finished unary call with code Unknown" error="`kustomize build <path to cached source>/components/my-app/staging` failed exit status 1: 2023/11/03 07:58:34 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied" grpc.code=Unknown grpc.method=GenerateManifest grpc.service=repository.RepoServerService grpc.start_time="2023-11-03T07:58:34Z" grpc.time_ms=650.84 span.kind=server system=grpc
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="symbolic reference 'HEAD' (refs/heads/main) resolved to 'a3fffe6133d7300a5b89daf639c7f13ab50ba9fc'"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=debug msg="cache hit for repo: https://gitlab.my-domain.com/my-repo.git revision: a3fffe6133d7300a5b89daf639c7f13ab50ba9fc"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:34Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=GetGitDirectories grpc.service=repository.RepoServerService grpc.start_time="2023-11-03T07:58:34Z" grpc.time_ms=1 span.kind=server system=grpc
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug msg="symbolic reference 'HEAD' (refs/heads/main) resolved to 'a3fffe6133d7300a5b89daf639c7f13ab50ba9fc'"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug msg="cache hit for repo: https://gitlab.my-domain.com/my-repo.git revision: a3fffe6133d7300a5b89daf639c7f13ab50ba9fc"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=GetGitDirectories grpc.service=repository.RepoServerService grpc.start_time="2023-11-03T07:58:35Z" grpc.time_ms=0.84 span.kind=server system=grpc
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug msg="symbolic reference 'HEAD' (refs/heads/main) resolved to 'a3fffe6133d7300a5b89daf639c7f13ab50ba9fc'"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug msg="Successfully restored read/write/execute permissions on /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=info msg="git fetch origin --tags --force --prune" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 execID=d90c0
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug duration=434.483567ms execID=d90c0
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=info msg=Trace args="[git fetch origin --tags --force --prune]" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 operation_name="exec git" time_ms=434.576298
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=info msg="git checkout --force a3fffe6133d7300a5b89daf639c7f13ab50ba9fc" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 execID=d662f
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug duration=4.473185ms execID=d662f
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=info msg=Trace args="[git checkout --force a3fffe6133d7300a5b89daf639c7f13ab50ba9fc]" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 operation_name="exec git" time_ms=4.560096
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=info msg="git clean -ffdx" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 execID=dcde1
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug duration=2.438548ms execID=dcde1
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=info msg=Trace args="[git clean -ffdx]" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 operation_name="exec git" time_ms=2.4981999999999998
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=info msg="git rev-parse HEAD" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 execID=e6c2c
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug msg="a3fffe6133d7300a5b89daf639c7f13ab50ba9fc\n" duration=2.044575ms execID=e6c2c
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=info msg=Trace args="[git rev-parse HEAD]" dir=/tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784 operation_name="exec git" time_ms=2.140686
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug msg="pluginSockFilePath is: /home/argocd/cmp-server/plugins" CWE=775 security=1
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug msg="No caCert found for repo gitlab.my-domain.com"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=info msg="kustomize build /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784/components/my-app/staging" dir= execID=57f93
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug duration=8.847466ms execID=57f93
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=error msg="`kustomize build /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784/components/my-app/staging` failed exit status 1: 2023/11/03 07:58:35 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied" execID=57f93
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=info msg=Trace args="[kustomize build /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784/components/my-app/staging]" dir= operation_name="exec kustomize" time_ms=8.977678
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug msg="getting manifests cache" appName=my-app-staging appSrc="{\"appSrc\":{\"repoURL\":\"\",\"path\":\"components/my-app/staging\"},\"srcRefs\":{}}" clusterInfo="1.22|<<edited>>" namespace=staging reason="GenerateManifests error" revision=a3fffe6133d7300a5b89daf639c7f13ab50ba9fc trackingKey="annotation:app.kubernetes.io/instance"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug msg="setting manifests cache" appName=my-app-staging appSrc="{\"appSrc\":{\"repoURL\":\"\",\"path\":\"components/my-app/staging\"},\"srcRefs\":{}}" clusterInfo="1.22|<<edited>>" namespace=staging reason="GenerateManifests error" revision=a3fffe6133d7300a5b89daf639c7f13ab50ba9fc trackingKey="annotation:app.kubernetes.io/instance"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=debug msg="Successfully removed permissions on /tmp/_argocd-repo/2cf8cca2-2367-44f9-b2c6-082368615784"
argocd-repo-server-648cfcc4cb-25w4r argocd-repo-server time="2023-11-03T07:58:35Z" level=error msg="finished unary call with code Unknown" error="`kustomize build <path to cached source>/components/my-app/staging` failed exit status 1: 2023/11/03 07:58:35 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied" grpc.code=Unknown grpc.method=GenerateManifest grpc.service=repository.RepoServerService grpc.start_time="2023-11-03T07:58:34Z" grpc.time_ms=588.899 span.kind=server system=grpc
```

**Comment by crenshaw-dev** [comment]
I'm so stumped.

Based on the logs given, I don't think it's a race condition with Argo CD's permission lockdown mechanism. I always see "Successfully restored permissions," <the error message>, and "Successfully removed permissions" in that order.

Here's what's weirder. Argo CD 2.8.4 runs Kustomize 5.1.0. The error message "unable to make loader at '.'" can only be produced by this function: https://github.com/kubernetes-sigs/kustomize/blob/kustomize/v5.1.0/api/loader/fileloader.go#L112C6-L112C24

That function is invoked by `kustomize add`, `kustomize create`, and `kustomize edit`. But we're getting the error from `kustomize build`.

So either I'm reading the Kustomize code very wrongly,  or something else weird is going on.

**Comment by crenshaw-dev** [comment]
Ignoring the fact that I can't find the problem code path, I have two bad theories:

1) `chmod` isn't atomic, and `kustomize build` is running before the fs changes actually take effect. I don't think that's the case, because `git checkout` works, and it certainly requires fs access.
2) `kustomize build` is running in the binary's path instead of the repo root, and there's something weird about the binary's path. We can test that theory with this PR: https://github.com/argoproj/argo-cd/pull/16229

Can anyone build/run that PR and see if it alleviates the problem?

**Comment by pepe9012** [comment]
I built ArgoCD v2.8.2 with the fix. So far no problems with kustomize. But after restarting the repo-server it always worked for some random time (5-120min). So I wouldn't give a final verdict until tomorrow.

**Comment by crenshaw-dev** [comment]
Even then I wouldn't 100% trust it, but it seems like at least a non-destructive change.

**Comment by pepe9012** [comment]
So far so good, if it lasts until Wednesday I will upgrade all instances and let you know if it is really fixed.

**Comment by crenshaw-dev** [comment]
Nice. I'll just need to fix the e2e tests, hopefully it's just a test problem and not a real problem.

**Comment by pepe9012** [comment]
@crenshaw-dev  After a few days, I don't see any problem with kustomize. I think you can safely integrate this fix into v2.8 and v2.9. Thanks a lot!

**Comment by praveen-livspace** [comment]
Thanks @pepe9012 for testing this out. Thanks @crenshaw-dev for debugging and fixing the issue. 

When is the PR expected to be merged?

**Comment by crenshaw-dev** [comment]
I think the e2e test failures might indicate an actual problem. I don't have time to tackle them now. Anyone else have time to take a look?

**Comment by prune998** [comment]
@crenshaw-dev I'm still experiencing the issue in v2.9.3+6eba5be (without your patch, which I'm going to test) and I'd like to have this PR moving forward so I'm not running on a custom build.
Not sure how to proceed as I obviously can't rebase your branch or do anything on the ArgoCD repo... WDYT ? 
Should I clone your branch and open another MR ? Or do you have some time to re-work on this one ? 
thx !

**Comment by prune998** [comment]
I cloned @crenshaw-dev PR to https://github.com/argoproj/argo-cd/pull/16652 and solved the e2e test.

**Comment by prune998** [comment]
So, even with the patch, the issue is still happening. 
Here's the logs of a `hard refresh` when the issue is triggered:

```


{
  "dir": "/tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7",
  "execID": "6a183",
  "level": "info",
  "msg": "git fetch origin --tags --force --prune",
  "time": "2023-12-21T16:07:01Z"
}
{
  "args": "[git fetch origin --tags --force --prune]",
  "dir": "/tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7",
  "level": "info",
  "msg": "Trace",
  "operation_name": "exec git",
  "time": "2023-12-21T16:07:02Z",
  "time_ms": 1370.861118
}
{
  "dir": "/tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7",
  "execID": "82856",
  "level": "info",
  "msg": "git checkout --force a4471a890ae6ccab3af9aad36aa05801934f23f3",
  "time": "2023-12-21T16:07:02Z"
}
{
  "args": "[git checkout --force a4471a890ae6ccab3af9aad36aa05801934f23f3]",
  "dir": "/tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7",
  "level": "info",
  "msg": "Trace",
  "operation_name": "exec git",
  "time": "2023-12-21T16:07:02Z",
  "time_ms": 9.378688
}
{
  "dir": "/tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7",
  "execID": "678c6",
  "level": "info",
  "msg": "git clean -ffdx",
  "time": "2023-12-21T16:07:02Z"
}
{
  "args": "[git clean -ffdx]",
  "dir": "/tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7",
  "level": "info",
  "msg": "Trace",
  "operation_name": "exec git",
  "time": "2023-12-21T16:07:02Z",
  "time_ms": 7.30541
}
{
  "dir": "/tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7",
  "execID": "04df7",
  "level": "info",
  "msg": "git rev-parse HEAD",
  "time": "2023-12-21T16:07:02Z"
}
{
  "dir": "/tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7",
  "execID": "83a51",
  "level": "info",
  "msg": "git rev-parse HEAD",
  "time": "2023-12-21T16:07:02Z"
}
{
  "args": "[git rev-parse HEAD]",
  "dir": "/tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7",
  "level": "info",
  "msg": "Trace",
  "operation_name": "exec git",
  "time": "2023-12-21T16:07:02Z",
  "time_ms": 5.916806
}
{
  "dir": "",
  "execID": "955dc",
  "level": "info",
  "msg": "kustomize build /tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7/deployments/k8s/overlays/qa/wk-qa-us-central-cluster",
  "time": "2023-12-21T16:07:02Z"
}
{
  "args": "[git rev-parse HEAD]",
  "dir": "/tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7",
  "level": "info",
  "msg": "Trace",
  "operation_name": "exec git",
  "time": "2023-12-21T16:07:02Z",
  "time_ms": 8.684288
}
{
  "dir": "",
  "execID": "91cda",
  "level": "info",
  "msg": "kustomize build /tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7/deployments/k8s/overlays/qa/wk-qa-us-central-cluster",
  "time": "2023-12-21T16:07:02Z"
}
{
  "execID": "955dc",
  "level": "error",
  "msg": "`kustomize build /tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7/deployments/k8s/overlays/qa/wk-qa-us-central-cluster` failed exit status 1: 2023/12/21 16:07:02 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied",
  "time": "2023-12-21T16:07:03Z"
}
{
  "args": "[kustomize build /tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7/deployments/k8s/overlays/qa/wk-qa-us-central-cluster]",
  "dir": "",
  "level": "info",
  "msg": "Trace",
  "operation_name": "exec kustomize",
  "time": "2023-12-21T16:07:03Z",
  "time_ms": 14.207365
}
{
  "error": "`kustomize build <path to cached source>/deployments/k8s/overlays/qa/wk-qa-us-central-cluster` failed exit status 1: 2023/12/21 16:07:02 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied",
  "grpc.code": "Unknown",
  "grpc.method": "GetAppDetails",
  "grpc.service": "repository.RepoServerService",
  "grpc.start_time": "2023-12-21T16:07:01Z",
  "grpc.time_ms": 1603.193,
  "level": "error",
  "msg": "finished unary call with code Unknown",
  "span.kind": "server",
  "system": "grpc",
  "time": "2023-12-21T16:07:03Z"
}
{
  "execID": "91cda",
  "level": "error",
  "msg": "`kustomize build /tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7/deployments/k8s/overlays/qa/wk-qa-us-central-cluster` failed exit status 1: 2023/12/21 16:07:03 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied",
  "time": "2023-12-21T16:07:03Z"
}
{
  "args": "[kustomize build /tmp/_argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7/deployments/k8s/overlays/qa/wk-qa-us-central-cluster]",
  "dir": "",
  "level": "info",
  "msg": "Trace",
  "operation_name": "exec kustomize",
  "time": "2023-12-21T16:07:03Z",
  "time_ms": 18.301377000000002
}
{
  "error": "`kustomize build <path to cached source>/deployments/k8s/overlays/qa/wk-qa-us-central-cluster` failed exit status 1: 2023/12/21 16:07:03 unable to make loader at '.'; not a valid directory: abs path error on '.' : stat .: permission denied",
  "grpc.code": "Unknown",
  "grpc.met
```

I exec into the pod, then:

```
k exec -ti argocd-repo-server-6dcd88bb46-jj6gc -c repo-server bash
cd /tmp
ls -l

d-wx------ 4 argocd argocd 4096 Dec 20 06:22 _argocd-repo
srwxr-xr-x 1 argocd argocd    0 Dec 20 06:21 reposerver-ask-pass.sock

cd _argocd-repo/
ls -l
ls: cannot open directory '.': Permission denied

cd ..
chmod u+r _argocd-repo/
ls -l _argocd-repo/

d--------- 5 argocd argocd 4096 Dec 20 06:22 0210738c-4ed1-4c05-83fc-0d78ecfa2023
d--------- 7 argocd argocd 4096 Dec 20 06:22 e539475b-5aac-46d5-8856-ff26b1b512d7

chmod u+rx _argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7

ls -l _argocd-repo/e539475b-5aac-46d5-8856-ff26b1b512d7/

-rw-r--r-- 1 argocd argocd 9551 Dec 20 06:22 README.md
drwxr-xr-x 3 argocd argocd 4096 Dec 20 06:22 cmd
drwxr-xr-x 4 argocd argocd 4096 Dec 20 06:22 deployments
drwxr-xr-x 3 argocd argocd 4096 Dec 20 06:22 docs
-rw-r--r-- 1 argocd argocd  117 Dec 20 06:22 renovate.json
```

After a `hard refresh`, the permissions are back to "not working".

I then tried running a `while true ; do chmod -R a+rwx _argocd-repo/ ; done` and do a hard refresh, and the error was gone.
So it seems that some of the commands used to `clone` the git repo are messing with permissions.

So it's really down to OS file permissions... I added a lot of debug in the `util/kustomize/kustomize.go` and `reposerver/repository/repository.go` to understand... my guess so far is something happening async and "locking out" one application, in the case where 2 apps use the same `repo/revision`... 
But it wouldn't explain why a `hard refresh` does not solve the solution once the repo is locked out.

Whatever I tested I wasn't able to reproduce the issue...
Let's wait and see if more debug gives a clue.

**Comment by obliadp** [comment]
This is happening to us quite frequently, and we have to restart all repo-server instances and do a hard refresh on apps.

Any update on this?

**Comment by prune998** [comment]
Well, it turn out @crenshaw-dev patch was working afterall, and my own patch was missing something. 

What is happening is that somehow (how ?) the clone on the project in the `/tmp/_argocd-repo/` folder is removed, so the `repo-server` can't run the `kustomize` command from the last repo's root folder.

The patch will be merged in 2.10 only, which should happen shortly, I hope :)

I corrected the patch + rebased + created a docker image for testing. You can use `prune/argocd:v2.9.3-prune` image to test the patch by using it in the `repo-server`. (assuming you're running 2.9 version).

**Comment by prune998** [comment]
To replicate the bug:

```
k exec -ti svc/argocd-repo-server bash

cd /tmp
chmod u+r _argocd-repo/
cd _argocd-repo/
ls -la

# remove one of the project's folder
chmod 0700 4048ada3-ecd6-4320-86a3-ee2887a68167/
rm -rf 4048ada3-ecd6-4320-86a3-ee2887a68167
```

With the bug, doing a `hard refresh` on the app associated with the deleted folder will log an error, while it will work OK with the patched version.

**Comment by obliadp** [comment]
Thanks for the update, looking forward to v2.10!

**Comment by yevon** [comment]
Just came across with this problem today, nice to see that this will be solved in 2.10

**Comment by yevon** [comment]
Any easy workarround without changing the docker image?

**Comment by prune998** [comment]
> Any easy workarround without changing the docker image?

nope, unless you want to automate a `repo-server` restart + a `hard refresh` of all your apps... :)

You can import my image into your own registry with

```
docker pull prune/argocd:v2.9.3-prune
docker tag prune/argocd:v2.9.3-prune <your-repo>:<your version>
docker push <your-repo>:<your version>
```

So you have more control. 
Or you can clone my PR and build it yourself :)

**Comment by glasser** [comment]
> Just came across with this problem today, nice to see that this will be solved in 2.10

It doesn't actually look yet like it will be solved in 2.10? The PR hasn't been merged. We are hitting this constantly on the one kustomize app we use.

**Comment by glasser** [comment]
@prune998 any interest in publishing a v2.10.0-rc3-prune? :)

**Comment by prune998** [comment]
> @prune998 any interest in publishing a v2.10.0-rc3-prune? :)

Will try either tomorrow or on monday :)

**Comment by prune998** [comment]
I just published a `prune/argocd:v2.10-rc4-prune` image. Feel free to test

**Comment by glasser** [comment]
Thanks, though is that with the last changes you made on the PR today?

**Comment by prune998** [comment]
it is without today's change, but the changes are just reverting back some defaults and will not change the behaviour... we're just trying to keep the compatibility with old behaviour while still patching the bug

**Comment by cilindrox** [comment]
Just a heads up that we run into this error on several apps after upgrading to `2.10.0` from `2.9.3`

**Comment by prune998** [comment]
> Just a heads up that we run into this error on several apps after upgrading to `2.10.0` from `2.9.3`

Can you try testing a custom image build with the patch ? 
`2.10.1` should be out soon, but I don't have control on the timeline

**Comment by cilindrox** [comment]
@prune998 I can do that - what's the tag?

**Comment by prune998** [comment]
It's merged in `master` and It seems it was cherry-picked in `release-2.10`

**Comment by cilindrox** [comment]
the latest `release-2.10` is what was failing though ^

**Comment by prune998** [comment]
> the latest `release-2.10` is what was failing though ^

maybe it was merged but no new image was generated ? In doubt, go with the `main`...

**Comment by glasser** [comment]
@cilindrox It's definitely not in v2.10.0 — what release notes make you think it is?  Waiting on v2.10.1 here.

**Comment by cilindrox** [comment]
my bad, was going by the automated release notes under that tag and I must have confused the issue ref. Going to try getting a build off `main` as suggested, thanks @glasser and @prune998

**Comment by rumstead** [comment]
The v2.10.1 release was cut and this change is in there

**Comment by cilindrox** [comment]
Quick update on this - after upgrading to v2.10.1 the issue seems fixed, however, I'm noticing after a while (~5h) the `application-controller` tends to hang and syncs/refreshes stop working. I've only been able to address this by restarting the instance.

I see the issue reflected in the logs with a deadline exceeded trace, eg:

```
time="2024-02-15T15:41:53Z" level=error msg="finished unary call with code Unknown" error="application refresh deadline exceeded" grpc.code=Unknown grpc.method=Get grpc.service=application.ApplicationService grpc.start_time="2024-02-15T15:40:53Z" grpc.time_ms=59999.84 span.kind=server system=grpc
```

**Comment by prune998** [comment]
thanks @cilindrox.
It could be related to the patch, or not... can you please open a new issue ? 
I had no time to upgrade yet... I will try to do it today.
